### PR TITLE
[LI-HOTFIX] Event based fetcher part 3/3: Adopting the event based fetcher model

### PR DESCRIPTION
**Pull request by gitlw**
This PR starts using the event based fetcher model, which replaces the existing lock based classes AbstractFetcherThread
and ReplicaFetcherThread.

To make implementation easier, I've decided to disable the following features for now
1. the JBOD logic implemented via ReplicaAlterLogDirsManager for now.
JBOD is not used within LinkedIn, so this change should be safe.

2. support for dynamically changing the fetcher thread pool size,
I checked the clusters in a given production fabric, including krc, queuing, metrics, venice, tracking, kac, data-deployment
and don't see a ZooKeeper override for the "num.replica.fetchers" config.

3. Many tests within the ConsumerBounceTest class are flaky, which is echod by the open source tickets
https://issues.apache.org/jira/browse/KAFKA-7893?jql=text%20~%20%22ConsumerBounceTest%22
Thus I'm disabling the entire test class for now.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by xiowu0** [comment]
would someone take down the broker by sending  "ALTER_REPLICA_LOG_DIRS" requests to the clusters?

File: core/src/main/scala/kafka/server/AbstractFetcherManager.scala [hunk 1]
```diff
@@ -188,49 +193,57 @@ abstract class AbstractFetcherManager[T <: AbstractFetcherThread](val name: Stri
...
   def shutdownIdleFetcherThreads(): Unit = {
     lock synchronized {
-      val keysToBeRemoved = new mutable.HashSet[BrokerIdAndFetcherId]
+      val futures = mutable.Map[BrokerIdAndFetcherId, KafkaFuture[Int]]()
+
       for ((key, fetcher) <- fetcherThreadMap) {
-        if (fetcher.idle) {
-          fetcher.shutdown()
-          keysToBeRemoved += key
+        futures.put(key, fetcher.getPartitionsCount())
```
Comment (xiowu0) [inline core/src/main/scala/kafka/server/AbstractFetcherManager.scala:229]: this function is called on every leaderAndIsr request as part of "becomeLeaderOrFollower". So how much delay can be introduced by sending "getPartitionsCount" request to event manager and waiting for result?

Comment (gitlw) [reply core/src/main/scala/kafka/server/AbstractFetcherManager.scala:229]: In upstream Kafka, the request handler thread needs to compete for the partitionMapLock in order to get the partitions count. And the fetcher.idle variable is a hotfix made internally by us. 
When switching to the event-based model, I feel it's cleaner to request the partition count using an event + future.
I'm not sure about the performance difference yet, but I have plans to eventually merge the remove partition, add partition and get partition count into one event if necessary. So ultimately the extra delay introduced by the GetPartitionCount event may not matter.

**Comment by gitlw** [comment]
> ALTER_REPLICA_LOG_DIRS

I've added a change in the latest commit to disallow the AlterReplicaLogDirs request in the authorization step,
which will make sure we are not vulnerable to this request.

File: core/src/test/scala/integration/kafka/server/DynamicBrokerReconfigurationTest.scala [hunk 2]
```diff
@@ -771,7 +772,7 @@ class DynamicBrokerReconfigurationTest extends ZooKeeperTestHarness with SaslSet
     verifyThreads("data-plane-kafka-socket-acceptor-", config.listeners.size)
 
     verifyProcessorMetrics()
-    verifyMarkPartitionsForTruncation()
+//    verifyMarkPartitionsForTruncation()
```
Comment (xiowu0) [inline core/src/test/scala/integration/kafka/server/DynamicBrokerReconfigurationTest.scala:775]: so this is only applicable when JBOD feature is used?

File: core/src/test/scala/unit/kafka/server/EventBasedFetcherTest.scala [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+import kafka.api.IntegrationTestHarness
+import org.apache.kafka.clients.producer.{KafkaProducer, ProducerConfig, ProducerRecord}
+import org.apache.kafka.common.TopicPartition
+import org.apache.kafka.common.serialization.ByteArraySerializer
+import org.junit.Test
+import java.util.Properties
+
+import kafka.utils.TestUtils
+
+class EventBasedFetcherTest extends IntegrationTestHarness{
```
Comment (xiowu0) [inline core/src/test/scala/unit/kafka/server/EventBasedFetcherTest.scala]: is there anything special in this test that covered the change in this patch?  If so, could you add some comments? Do we have other existing tests that cover the basic send function?

File: core/src/test/scala/integration/kafka/server/DynamicBrokerReconfigurationTest.scala (hunk 2 above)
Comment (gitlw) [reply core/src/test/scala/integration/kafka/server/DynamicBrokerReconfigurationTest.scala:775]: yes

File: core/src/test/scala/unit/kafka/server/EventBasedFetcherTest.scala (hunk 3 above)
Comment (gitlw) [reply core/src/test/scala/unit/kafka/server/EventBasedFetcherTest.scala]: No, there isn't anything special. So I'll simply remove this test.

**Review by xiowu0 — APPROVED** [review]
LGTM.  

Do you plan to re-enable JBOD in the future?

**Comment by gitlw** [comment]
@xiowu0 Thanks for the review. I'll revisit JBOD only if it turns out to be needed within LI.

**Pull request merged**